shell/client/grain-client.js

```diff
@@ -2024,3 +2025,112 @@ Router.map(function () {
     },
   });
 });
+
+Meteor.startup(function () {
+  if (isStandalone()) {
+    // TODO(soon): delete all other routes (maybe change how they're created?)
+    const route = Router.routes.root;
+    _.extend(route.options, {
+      template: "grain",
+      loadingTemplate: "loadingNoMessage",
+
+      waitOn: function () {
+        const standalone = globalDb.collections.standaloneDomains.findOne({
+          _id: window.location.hostname, });
+        return [
+          Meteor.subscribe("devPackages"),
+          Meteor.subscribe("tokenInfo", standalone && standalone.token),
+          Meteor.subscribe("standaloneDomain", window.location.hostname),
+          Meteor.subscribe("grainsMenu"),
+          // This subscription gives us the data we need for deciding whether to automatically reveal
+          // our identity.
+          // TODO(soon): Subscribe to contacts instead.
+        ];
+      },
+
+      onBeforeAction: function () {
+        // Don't do anything for non-account users.
+        if (Meteor.userId() && !Meteor.user().loginIdentities) return;
+
+        // Only run the hook once.
+        if (this.state.get("beforeActionHookRan")) return this.next();
+        this.state.set("beforeActionHookRan", true);
+
+        Tracker.nonreactive(() => {
+          const token = globalDb.collections.standaloneDomains.findOne({
+            _id: window.location.hostname, }).token;
+          const path = "/" + (this.params.path || "") + (this.originalUrl.match(/[#?].*$/) || "");
+          const hash = this.params.hash;
+
+          const tokenInfo = TokenInfo.findOne({ _id: token });
+          if (!tokenInfo) {
```

Comment:
There's a lot of code here that is copy/pasted from elsewhere. We should factor it out so that it can be shared.